scripts/powershell/sub-cleanup/sub-cleanup.ps1

```diff
@@ -0,0 +1,111 @@
+#!/usr/bin/env pwsh -c
+
+<#
+.SYNOPSIS
+Subscription cleanup script
+
+.DESCRIPTION
+Cleans up non-compliant resource groups in a specified Azure
+subscription.  Complient resource groups are named with a valid
+Microsoft alias or using the format: alias-<identifier>
+
+.NOTES
+You can add the tag CleanupException=true to the resource group
+to have it skipped as an exception. This can be done with CLI
+or in the portal.
+
+Requires PowerShell module Az and RemoteSigned excecution policy.
+Commands to make this happen:
+  PS> Install-Module -Name Az -AllowClobber
+  PS> Set-ExecutionPolicy RemoteSigned
+
+.EXAMPLE
+PS> sub-cleanup.ps1 -Subscription {guid} -Audit false
+
+.PARAMETER Subscription
+Subscription ID (guid) for the subscription to be cleaned.
+
+.PARAMETER Audit
+When specified, deletion operations are just logged, but not performed.
+#>
+
+param(
+  [Parameter(Mandatory = $true)]
+  [string]$Subscription,
+  
+  [Parameter(Mandatory = $false)]
+  [boolean]$Audit=$true
+)
+
+# Check if we're in audit mode (no actual deletions)
+if ($Audit)
+{
+    Write-Host -ForegroundColor Green "Running in audit mode. Nothing will be deleted."
+}
+else
+{
+    Write-Host -ForegroundColor Red "!!!NOT IN AUDIT MODE. RESOURCE GROUPS WILL BE DELETED!!!"
+}
+
+# Connect and select active subscription
+Connect-AzAccount
+Select-AzSubscription -Subscription $Subscription
+
+#initialize lists
+$deleted = @()
+$skipped = @()
+
+# Loop through the resource groups and delete any with non-compliant names
+foreach ($resourceGroup in Get-AzResourceGroup | Sort ResourceGroupName)
+{
+    # skip exceptions
+    if ($resourceGroup.Tags -and $resourceGroup.Tags["CleanupException"] -eq "true")
+    {
+        Write-Host " Skipping tagged exception resource group: $($resourceGroup.ResourceGroupName)"
+        $skipped += $($resourceGroup.ResourceGroupName)
+        continue
+    }
+
+    # check compliance (formatting first, then validate alias) and skip if compliant
+    if ($resourceGroup.ResourceGroupName -match "^(rg-)?((t-|a-|v-)?[a-z,A-Z]{3,15})([-_]{1}.*)?$" -and (Get-AzAdUser -UserPrincipalName "$($Matches.2)@microsoft.com"))
+    {
+        Write-Host " Skipping compliant resource group: $($resourceGroup.ResourceGroupName)"
+        $skipped += $($resourceGroup.ResourceGroupName)
+        continue
+    }
+
+    Write-Host -ForegroundColor Red -NoNewline "Deleting non-compliant resource group: $($resourceGroup.ResourceGroupName)"
+    
+    # delete
+    $deleted += $($resourceGroup.ResourceGroupName)
+    if ($Audit)
+    {
+        Write-Host " Audit:Skipped."
+    }
+    else
+    {
+        if (Remove-AzResourceGroup -Name $resourceGroup.ResourceGroupName -Force)
```

Comment:
For example, if you use:
```powershell
[CmdletBinding(SupportsConfirm=$true)]
param (
  # ...
)
```

That should declare a standard `-WhatIf` which then `Remove-AzResourceGroup` should use (the preference variable) automatically, So then `Remove-AzResourceGroup` prompts to actually remove or, if in `-WhatIf` mode, shows what it would do but not actually do it.

Reply:
Turns out it's not so straight forward for a couple of reasons, including that `Remove-AzResourceGroup` does not use standard `ShouldProcess` behavior. Also turns out that selecting a different sub supports `-WhatIf`, so if you pass `-WhatIf` to the script, the sub isn't actually switched and you won't enumerate the resource groups you thought you were!

I worked through the problems on my own personal MSA account (and very carefully I might add) and came up with this if you want to use it as a starting point: https://gist.github.com/heaths/a3810f36a56177ae1cbe4cf57af0b02f

If you wanted, you could replace all the `Write-Verbose` lines with `Write-Host` or even [`Log`](https://github.com/Azure/azure-sdk-tools/blob/7015d6ab0c85338831298f067eb32328874aa1dc/eng/common/TestResources/New-TestResources.ps1#L87-L89) method the TestResources scripts use for easier automation diagnostics in Azure Pipelines.